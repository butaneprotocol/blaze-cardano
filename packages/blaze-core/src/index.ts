export * from "./types";
export * from "./util";
export * from "./params";
export * from "./pretty";
export * from "./errors";
