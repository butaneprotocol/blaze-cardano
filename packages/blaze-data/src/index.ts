import {
  Kind,
  Type,
  PatternStringExact,
  PatternNumberExact,
  type Static,
  type TEnum,
  type TSchema,
  type TObject,
  type TNumber,
  type TBigInt,
  type TString,
  type TArray,
  type TUnion,
  type TLiteral,
  type TRecord,
  type TBoolean,
  type TRef,
  type TThis,
} from "@sinclair/typebox";
import {
  ConstrPlutusData,
  fromHex,
  HexBlob,
  PlutusData,
  PlutusList,
  PlutusMap,
  toHex,
} from "@blaze-cardano/core";
export { Type, type TSchema, type TArray } from "@sinclair/typebox";

export const Void = (): PlutusData =>
  PlutusData.newConstrPlutusData(new ConstrPlutusData(0n, new PlutusList()));

export const TPlutusData: TSchema = Type.Unsafe<PlutusData>(Type.Any());

export type Exact<T> = T extends TSchema ? Static<T> : T;

const isArray = (t: TSchema): t is TArray => t[Kind] === "Array";
const isBigInt = (t: TSchema): t is TBigInt => t[Kind] === "BigInt";
const isBoolean = (t: TSchema): t is TBoolean => t[Kind] === "Boolean";
const isLiteral = (t: TSchema): t is TLiteral => t[Kind] === "Literal";
const isNumber = (t: TSchema): t is TNumber => t[Kind] === "Number";
const isObject = (t: TSchema): t is TObject => t[Kind] === "Object";
const isRecord = (t: TSchema): t is TRecord => t[Kind] === "Record";
const isRef = (t: TSchema): t is TRef => t[Kind] === "Ref";
const isString = (t: TSchema): t is TString => t[Kind] === "String";
const isThis = (t: TSchema): t is TThis => t[Kind] === "This";
const isUnion = (t: TSchema): t is TUnion => t[Kind] === "Union";

export function serialize<T extends TSchema>(
  type: T,
  data: Exact<T>,
): PlutusData {
  const result = _serialize(type, data, ["root"]);
  return result;
}

export function _serialize<T extends TSchema>(
  type: T,
  data: Exact<T>,
  path: string[],
  defs?: Record<string, TSchema>,
): PlutusData {
  if (!type) {
    throw new Error(
      `Cannot serialize ${data} without a type. Found at ${path.join(".")}`,
    );
  }

  if (data instanceof PlutusData) {
    return data;
  }

  if ("$ref" in type) {
    defs ??= type["$defs"];
    if (!defs) {
      throw new Error(
        `Could not resolve type ${type["$ref"]} at ${path.join(".")} because we don't have a definitions dictionary`,
      );
    }

    const resolvedType = defs[type["$ref"]];
    if (!resolvedType) {
      throw new Error(
        `Could not resolve type ${type["$ref"]} at ${path.join(".")}`,
      );
    }
    return _serialize(resolvedType, data, path, defs);
  }

  if ("anyOf" in type) {
    const enumType = type as unknown as TEnum;
    if (enumType.anyOf.length === 1) {
      return _serialize(type["anyOf"][0]! as T, data, path, defs);
    }
  }

  if ("type" in type) {
    switch (type["type"]) {
      case "bigint":
        {
          if (data instanceof BigInt || typeof data === "bigint") {
            return PlutusData.newInteger(data as bigint);
          } else if (typeof data === "number") {
            return PlutusData.newInteger(BigInt(data));
          } else if (typeof data === "string") {
            return PlutusData.newInteger(BigInt(data));
          }
          throw new Error(
            `Invalid data for plutus integer at path ${path.join(".")}: ${typeof data}`,
          );
        }
        break;
      case "string":
        {
          return PlutusData.newBytes(fromHex(data as string));
        }
        break;
      case "array":
        {
          // Handle tuple-enums
          if ("ctor" in type) {
            if (
              typeof type["ctor"] !== "bigint" &&
              typeof type["ctor"] !== "number"
            ) {
              throw new Error(
                `Invalid object at ${path.join(".")}: Enum variant constructor index must be a number`,
              );
            }

            const constructor = BigInt(type["ctor"]);
            const fieldData = new PlutusList();

            const array = data as Array<any>;
            for (let idx = 0; idx < array.length; idx++) {
              let fieldType: any = undefined;
              if (Array.isArray(type["items"])) {
                // A tuple will be a schema with an array of item types
                fieldType = type["items"][idx];
              } else {
                // Otherwise it's just a homogeneous array, reusing the same type each time
                fieldType = type["items"];
              }
              fieldData.add(
                _serialize(
                  fieldType,
                  array[idx],
                  [...path, idx.toString()],
                  defs,
                ),
              );
            }

            return PlutusData.newConstrPlutusData(
              new ConstrPlutusData(constructor, fieldData),
            );
          } else {
            const items = new PlutusList();
            const array = data as Array<any>;
            for (let idx = 0; idx < array.length; idx++) {
              let itemType: any = undefined;
              if (Array.isArray(type["items"])) {
                // A tuple will be a schema with an array of item types
                itemType = type["items"][idx];
              } else {
                // Otherwise it's just a homogeneous array, reusing the same type each time
                itemType = type["items"];
              }
              items.add(
                _serialize(
                  itemType,
                  array[idx],
                  [...path, idx.toString()],
                  defs,
                ),
              );
            }
            return PlutusData.newList(items);
          }
        }
        break;
      case "object":
        {
          // Deal with maps
          if ("patternProperties" in type) {
            const valueType = type["patternProperties"]["^(.*)$"];
            const dataMap = new PlutusMap();
            const object = data as Record<string, any>;
            for (const key in object) {
              const plutusKey = PlutusData.newBytes(fromHex(key));
              const plutusValue = _serialize(
                valueType,
                object[key],
                [...path, key],
                defs,
              );
              dataMap.insert(plutusKey, plutusValue);
            }
            return PlutusData.newMap(dataMap);
          }
          if (typeof type["ctor"] === "undefined") {
            throw new Error(
              `Invalid object at ${path.join(".")}: Enum variant must have a constructor index`,
            );
          }
          if (
            typeof type["ctor"] !== "bigint" &&
            typeof type["ctor"] !== "number"
          ) {
            throw new Error(
              `Invalid object at ${path.join(".")}: Enum variant constructor index must be a number`,
            );
          }
          const constructor = BigInt(type["ctor"]);
          const object = data as Record<string, any>;
          const fields = type["properties"] as Record<string, T>;
          const fieldData = new PlutusList();
          for (const [field, fieldType] of Object.entries(fields)) {
            fieldData.add(
              _serialize(fieldType, object[field], [...path, field], defs),
            );
          }
          return PlutusData.newConstrPlutusData(
            new ConstrPlutusData(constructor, fieldData),
          );
        }
        break;
      default: {
        console.log("Unhandled: ", type["type"]);
      }
    }
  } else if ("anyOf" in type) {
    // This is a plutus data constructor
    if (type["anyOf"].length === 1) {
      // Just one variant, so we can directly serialize it
      return _serialize(type["anyOf"][0], data, path, defs);
    }
    // Otherwise, it could be either boolean or optional
    // Finally, we should try to identify which variant this is:
    switch (typeof data) {
      case "string": {
        if (!/[A-Z]/.test(data[0]!)) {
          throw new Error(
            `Invalid enum variant at ${path.join(".")}: Enum variants must start with an uppercase letter`,
          );
        }

        const variant = type["anyOf"].find(
          (variant: TSchema) =>
            variant["type"] === "string" && variant["const"] === data,
        );
        if (!variant) {
          throw new Error(
            `Invalid enum variant at ${path.join(".")}: Enum variant not found`,
          );
        }
        if (typeof variant["ctor"] === "undefined") {
          throw new Error(
            `Invalid enum variant at ${path.join(".")}: Enum variant constructor index must be defined`,
          );
        }
        if (
          typeof variant["ctor"] !== "bigint" &&
          typeof variant["ctor"] !== "number"
        ) {
          throw new Error(
            `Invalid enum variant at ${path.join(".")}: Enum variant constructor index must be a number`,
          );
        }

        return PlutusData.newConstrPlutusData(
          new ConstrPlutusData(BigInt(variant["ctor"]), new PlutusList()),
        );
      }
      case "object": {
        if (typeof data === "undefined" || data === null) {
          throw new Error(
            `Invalid object at ${path.join(".")}: Enum variants cannot be empty`,
          );
        }
        const variantTitle = Object.keys(data)[0];
        if (!variantTitle) {
          throw new Error(
            `Invalid object at ${path.join(".")}: Enum has no keys`,
          );
        }
        if (!/^[A-Z]/.test(variantTitle)) {
          throw new Error(
            `Invalid object at ${path.join(".")}: Enum variants must start with an uppercase letter`,
          );
        }
        // Find which variant this is
        const variant = type["anyOf"].find(
          (variant: TSchema) =>
            variant["type"] === "object" &&
            Object.keys(variant["properties"])[0] === variantTitle,
        );
        if (!variant) {
          throw new Error(
            `Invalid object at ${path.join(".")}: Enum variant not found`,
          );
        }
        const schema = variant["properties"][variantTitle];
        return _serialize(schema, (data as any)[variantTitle], [
          ...path,
          variantTitle,
        ]);
      }
    }
  }

  return PlutusData.fromCbor(HexBlob("01"));
}

export function parse<T extends TSchema>(
  type: T,
  data: PlutusData,
  defs?: Record<string, TSchema>,
): Exact<T> {
  return _parse(type, data, ["root"], { ...defs });
}

export function _parse<T extends TSchema>(
  type: T,
  data: PlutusData,
  path: string[],
  defs: Record<string, TSchema>,
): Exact<T> {
  if (type.$id) {
    defs[type.$id] = type;
  }
  type = resolveType<T>(type, path, defs);
  if (isRef(type) || isThis(type)) {
    const realType = defs[type.$ref];
    if (!realType) {
      throw new Error(
        `Invalid type at ${path.join(".")}: Unrecognized reference ${type.$ref}`,
      );
    }
    return _parse(realType, data, path, defs) as Exact<T>;
  }
  if (isBigInt(type)) {
    const value = data.asInteger();
    if (value === undefined) {
      throw new Error(
        `Invalid type at ${path.join(".")}: Expected integer, got ${data.getKind()}.`,
      );
    }
    return value as Exact<T>;
  }
  if (isNumber(type)) {
    const value = data.asInteger();
    if (value === undefined) {
      throw new Error(
        `Invalid type at ${path.join(".")}: Expected integer, got ${data.getKind()}.`,
      );
    }
    if (value < Number.MIN_SAFE_INTEGER || value > Number.MAX_SAFE_INTEGER) {
      throw new Error(`Invalid number at ${path.join(".")}: out of range`);
    }
    return Number(value) as Exact<T>;
  }
  if (isString(type)) {
    const value = data.asBoundedBytes();
    if (!value) {
      throw new Error(
        `Invalid type at ${path.join(".")}: Expected bytes, got ${data.getKind()}.`,
      );
    }
    return toHex(value) as Exact<T>;
  }
  if (isBoolean(type)) {
    return _parse(BOOLEAN_ENUM_SCHEMA, data, path, defs) as Exact<T>;
  }
  if (isLiteral(type)) {
    const value = data.asConstrPlutusData();
    if (!value) {
      throw new Error(
        `Invalid type at ${path.join(".")}: Expected constr, got ${data.getKind()}.`,
      );
    }
    if ("ctor" in type) {
      const ctor = BigInt(type["ctor"]);
      if (ctor !== value.getAlternative()) {
        throw new Error(
          `Invalid literal at ${path.join(".")}: Expected alternative ${ctor}, got ${value.getAlternative()}.`,
        );
      }
    } else {
      throw new Error(
        `Invalid literal at ${path.join(".")}: Enum variant not found.`,
      );
    }
    if (value.getData().getLength()) {
      throw new Error(
        `Invalid literal at ${path.join(".")}: Enum variants are not expected to have fields.`,
      );
    }
    return type.const as Exact<T>;
  }
  if (isRecord(type)) {
    const value = data.asMap();
    if (!value) {
      throw new Error(
        `Invalid type at ${path.join(".")}: Expected map, got ${data.getKind()}.`,
      );
    }
    const result: Record<any, any> = {};
    const keys = value.getKeys();
    for (let i = 0; i < value.getLength(); ++i) {
      const key = keys.get(i);
      const item = value.get(key)!;
      if (key.asInteger() !== undefined) {
        const intKey = Number(key.asInteger());
        const itemType = type.patternProperties[PatternNumberExact];
        if (!itemType) {
          throw new Error(
            `Invalid map at ${path.join(".")}: Unexpected numeric key.`,
          );
        }
        result[intKey] = _parse(
          itemType,
          item,
          [...path, intKey.toString()],
          defs,
        );
      } else if (key.asBoundedBytes()) {
        const bytesKey = toHex(key.asBoundedBytes()!);
        const itemType = type.patternProperties[PatternStringExact];
        if (!itemType) {
          throw new Error(
            `Invalid map at ${path.join(".")}: Unexpected non-numeric key.`,
          );
        }
        result[bytesKey] = _parse(itemType, item, [...path, bytesKey], defs);
      } else {
        throw new Error(
          `Invalid map at ${path.join(".")}: Keys of type ${key.getKind()} are not supported.`,
        );
      }
    }
    return result as Exact<T>;
  }
  if (isArray(type)) {
    const value = data.asList();
    if (!value) {
      throw new Error(
        `Invalid type at ${path.join(".")}: Expected list, got ${data.getKind()}.`,
      );
    }
    const result = [];
    for (let i = 0; i < value.getLength(); ++i) {
      const item = value.get(i);
      result.push(_parse(type.items, item, [...path, i.toString()], defs));
    }
    return result as Exact<T>;
  }
  if (isObject(type)) {
    const value = data.asConstrPlutusData();
    if (!value) {
      throw new Error(
        `Invalid type at ${path.join(".")}: Expected constr, got ${data.getKind()}.`,
      );
    }
    if ("ctor" in type) {
      const ctor = BigInt(type["ctor"]);
      if (ctor !== value.getAlternative()) {
        throw new Error(
          `Invalid constr at ${path.join(".")}: Expected alternative ${ctor}, got ${value.getAlternative()}.`,
        );
      }
    }
    const result: Record<string, unknown> = {};

    const fields = value.getData();
    let fieldIndex = 0;
    for (const [fieldName, fieldType] of Object.entries(type.properties)) {
      const fieldData = fields.get(fieldIndex);
      if (!fieldData) {
        const required = (type.required?.indexOf(fieldName) ?? -1) >= 0;
        if (required && !fieldData) {
          throw new Error(
            `Invalid object at ${path.join(".")}: Missing required field ${fieldName}`,
          );
        }
      } else {
        result[fieldName] = _parse(
          fieldType,
          fieldData,
          [...path, fieldName],
          defs,
        );
      }
      ++fieldIndex;
    }
    return result as Exact<T>;
  }
  if (isUnion(type)) {
    const value = data.asConstrPlutusData();
    if (!value) {
      throw new Error(
        `Invalid type at ${path.join(".")}: Expected constr, got ${data.getKind()}.`,
      );
    }
    const actualCtor = value.getAlternative();
    const variantIndex = type.anyOf.findIndex((s) => {
      const resolved = resolveType(s, path, defs);
      return "ctor" in resolved && BigInt(resolved["ctor"]) === actualCtor;
    });
    if (variantIndex < 0) {
      throw new Error(
        `Invalid union at ${path.join(".")}: Unrecognized variant ${actualCtor}.`,
      );
    }
    return _parse(
      type.anyOf[variantIndex]!,
      data,
      [...path, variantIndex.toString()],
      defs,
    ) as Exact<T>;
  }
  throw new Error(
    `Invalid type at ${path.join(".")}: Unrecognized type "${type[Kind]}".`,
  );
}

function resolveType<T extends TSchema>(
  type: T,
  path: string[],
  defs: Record<string, TSchema>,
): T {
  if (isRef(type) || isThis(type)) {
    const realType = defs[type.$ref];
    if (!realType) {
      throw new Error(
        `Invalid type at ${path.join(".")}: Unrecognized reference ${type.$ref}`,
      );
    }
    return resolveType(realType, path, defs) as T;
  }
  return type;
}

const BOOLEAN_ENUM_SCHEMA = Type.Union([
  Type.Literal(false, { ctor: 0 }),
  Type.Literal(true, { ctor: 1 }),
]);
