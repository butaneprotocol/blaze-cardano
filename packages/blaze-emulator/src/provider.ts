import type {
  Address,
  ProtocolParameters,
  AssetId,
  TransactionId,
  Transaction,
  Redeemers,
  DatumHash,
} from "@blaze-cardano/core";
import {
  TransactionInput,
  PlutusData,
  TransactionOutput,
  NetworkId,
} from "@blaze-cardano/core";
import { TransactionUnspentOutput } from "@blaze-cardano/core";
import { Provider } from "@blaze-cardano/query";
import type { Emulator } from "./emulator";

/**
 * The EmulatorProvider class implements the Provider interface.
 * It provides methods to interact with the Emulator.
 */
export class EmulatorProvider extends Provider {
  /**
   * The Emulator instance.
   */
  private emulator: Emulator;

  constructor(emulator: Emulator) {
    super(NetworkId.Testnet);
    this.emulator = emulator;
  }
  getParameters(): Promise<ProtocolParameters> {
    return Promise.resolve(this.emulator.params);
  }

  getUnspentOutputs(address: Address): Promise<TransactionUnspentOutput[]> {
    const utxos: TransactionUnspentOutput[] = [];
    const addressBytes = address.toBytes();
    for (const utxo of this.emulator.utxos()) {
      if (utxo.output().address().toBytes() == addressBytes) {
        utxos.push(TransactionUnspentOutput.fromCore(utxo.toCore()));
      }
    }
    return Promise.resolve(utxos);
  }

  getUnspentOutputsWithAsset(
    address: Address,
    unit: AssetId,
  ): Promise<TransactionUnspentOutput[]> {
    const utxos: TransactionUnspentOutput[] = [];
    const addressBytes = address.toBytes();
    for (const utxo of this.emulator.utxos()) {
      if (
        utxo.output().address().toBytes() == addressBytes &&
        utxo.output().amount().multiasset()?.get(unit) !== undefined
      ) {
        utxos.push(TransactionUnspentOutput.fromCore(utxo.toCore()));
      }
    }
    return Promise.resolve(utxos);
  }

  getUnspentOutputByNFT(unit: AssetId): Promise<TransactionUnspentOutput> {
    for (const utxo of this.emulator.utxos()) {
      if (utxo.output().amount().multiasset()?.get(unit) != undefined) {
        return Promise.resolve(
          TransactionUnspentOutput.fromCore(utxo.toCore()),
        );
      }
    }
    return Promise.reject(
      "getUnspentOutputByNFT: emulated ledger had no UTxO with NFT",
    );
  }

  resolveUnspentOutputs(
    txIns: TransactionInput[],
  ): Promise<TransactionUnspentOutput[]> {
    const utxos = [];
    for (const txIn of txIns) {
      const out = this.emulator.getOutput(txIn);
      if (out)
        utxos.push(
          new TransactionUnspentOutput(
            TransactionInput.fromCore(txIn.toCore()),
            TransactionOutput.fromCore(out.toCore()),
          ),
        );
    }
    return Promise.resolve(utxos);
  }

  resolveDatum(datumHash: DatumHash): Promise<PlutusData> {
    return Promise.resolve(
      PlutusData.fromCore(this.emulator.datumHashes[datumHash]!.toCore()),
    );
  }

  awaitTransactionConfirmation(
    txId: TransactionId,
    _timeout?: number | undefined,
  ): Promise<boolean> {
    this.emulator.awaitTransactionConfirmation(txId);
    return Promise.resolve(true);
  }

  postTransactionToChain(tx: Transaction): Promise<TransactionId> {
    return this.emulator.submitTransaction(tx);
  }

  evaluateTransaction(
    tx: Transaction,
    additionalUtxos: TransactionUnspentOutput[],
  ): Promise<Redeemers> {
    return this.emulator.evaluator(tx, additionalUtxos);
  }
}
