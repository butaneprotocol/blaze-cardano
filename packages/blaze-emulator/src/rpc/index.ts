export { default as app } from "./app";
export type { AppType } from "./app";
export { startRpcServer } from "./server";
export { createEmulatorRpcClient } from "./sdk/client";
export type { EmulatorRpcClient } from "./sdk/client";
