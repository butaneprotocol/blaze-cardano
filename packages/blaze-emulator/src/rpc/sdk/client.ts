import { hc } from "hono/client";
import type app from "../app";

export type EmulatorRpcClient = ReturnType<typeof hc<typeof app>>;

export const createEmulatorRpcClient = (baseUrl: string): EmulatorRpcClient =>
  hc<typeof app>(baseUrl);
