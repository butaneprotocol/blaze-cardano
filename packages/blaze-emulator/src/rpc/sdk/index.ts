export { createEmulatorRpcClient } from './client';
export type { EmulatorRpcClient } from './client';
