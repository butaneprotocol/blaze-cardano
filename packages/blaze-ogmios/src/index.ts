export { Schema } from "./schema";
export * as Unwrapped from "./unwrapped";
export * from "./wrapped";
