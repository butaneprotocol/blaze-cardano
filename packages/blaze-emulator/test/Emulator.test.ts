import type { Bip32PrivateKeyHex, Address } from "@blaze-cardano/core";
import {
  NetworkId,
  TransactionInput,
  TransactionId,
  PolicyId,
  AssetName,
  AssetId,
  addressFromCredential,
  Credential,
  CredentialType,
  TransactionUnspentOutput,
  RewardAccount,
} from "@blaze-cardano/core";
import { HotWallet } from "@blaze-cardano/wallet";
import { Emulator, EmulatorProvider } from "../src";
import {
  ONE_PLUTUS_DATA,
  VOID_PLUTUS_DATA,
  alwaysTrueScript,
  generateAccount,
  generateGenesisOutputs,
  signAndSubmit,
} from "./util";
import { Blaze, makeValue } from "@blaze-cardano/sdk";

function isDefined<T>(value: T | undefined): asserts value is T {
  expect(value).toBeDefined();
}

describe("Emulator", () => {
  let emulator: Emulator;
  let masterkeyHex1: Bip32PrivateKeyHex;
  let address1: Address;
  let wallet1: HotWallet;
  let masterkeyHex2: Bip32PrivateKeyHex;
  let address2: Address;
  let wallet2: HotWallet;
  let blaze: Blaze<EmulatorProvider, HotWallet>;
  let provider: EmulatorProvider;

  beforeAll(async () => {
    ({ address: address1, masterkeyHex: masterkeyHex1 } =
      await generateAccount());
    ({ address: address2, masterkeyHex: masterkeyHex2 } =
      await generateAccount());
  });

  beforeEach(async () => {
    emulator = new Emulator([
      ...generateGenesisOutputs(address1),
      ...generateGenesisOutputs(address2),
    ]);
    provider = new EmulatorProvider(emulator);
    wallet1 = await HotWallet.fromMasterkey(masterkeyHex1, provider);
    wallet2 = await HotWallet.fromMasterkey(masterkeyHex2, provider);
    blaze = await Blaze.from(provider, wallet1);
  });

  test("Should be able to get a genesis UTxO", async () => {
    const inp = new TransactionInput(TransactionId("00".repeat(32)), 0n);
    const out = emulator.getOutput(inp);
    isDefined(out);
    expect(out.address()).toEqual<Address>(wallet1.address);
    expect(out?.amount().coin()).toEqual(1_000_000_000n);
  });

  test("Should be able to pay from one wallet to another", async () => {
    const tx = await blaze
      .newTransaction()
      .payLovelace(wallet2.address, 2_000_000_000n)
      .complete();
    const txHash = await signAndSubmit(tx, blaze);
    emulator.awaitTransactionConfirmation(txHash);
    const out = emulator.getOutput(new TransactionInput(txHash, 0n));
    isDefined(out);
    expect(out.address()).toEqual(wallet2.address);
    expect(out.amount().coin()).toEqual(2_000_000_000n);
  });

  test("Should be able to spend from a script", async () => {
    const tx = await blaze
      .newTransaction()
      .lockAssets(
        addressFromCredential(
          NetworkId.Testnet,
          Credential.fromCore({
            type: CredentialType.ScriptHash,
            hash: alwaysTrueScript.hash(),
          }),
        ),
        makeValue(1_000_000_000n),
        ONE_PLUTUS_DATA,
      )
      .complete();
    const txHash = await signAndSubmit(tx, blaze);
    emulator.awaitTransactionConfirmation(txHash);
    const inp = new TransactionInput(txHash, 0n);
    const out = emulator.getOutput(inp);
    isDefined(out);
    isDefined(out.datum());
    const spendTx = await blaze
      .newTransaction()
      .addInput(new TransactionUnspentOutput(inp, out), VOID_PLUTUS_DATA)
      .lockAssets(
        addressFromCredential(
          NetworkId.Testnet,
          Credential.fromCore({
            type: CredentialType.ScriptHash,
            hash: alwaysTrueScript.hash(),
          }),
        ),
        makeValue(1_000_000_000n),
        ONE_PLUTUS_DATA,
      )
      .provideScript(alwaysTrueScript)
      .complete();
    const spendTxHash = await signAndSubmit(spendTx, blaze);
    emulator.awaitTransactionConfirmation(spendTxHash);
    const out2 = emulator.getOutput(new TransactionInput(spendTxHash, 0n));
    isDefined(out2);
    expect(out2.address()).toEqual(wallet1.address);
  });

  test("Should be able to spend from a script with a reference input", async () => {
    const refTx = await blaze
      .newTransaction()
      .deployScript(alwaysTrueScript)
      .complete();
    const refTxHash = await signAndSubmit(refTx, blaze);
    emulator.awaitTransactionConfirmation(refTxHash);
    const refUtxo = await provider.resolveScriptRef(alwaysTrueScript);
    isDefined(refUtxo);
    // const refIn = new TransactionInput(refTxHash, 0n);
    // const refUtxo = new TransactionUnspentOutput(
    //   refIn,
    //   emulator.getOutput(refIn)!
    // );

    const tx = await blaze
      .newTransaction()
      .lockAssets(
        addressFromCredential(
          NetworkId.Testnet,
          Credential.fromCore({
            type: CredentialType.ScriptHash,
            hash: alwaysTrueScript.hash(),
          }),
        ),
        makeValue(1_000_000_000n),
        ONE_PLUTUS_DATA,
      )
      .complete();

    const txHash = await signAndSubmit(tx, blaze);
    emulator.awaitTransactionConfirmation(txHash);
    const inp = new TransactionInput(txHash, 0n);
    const out = emulator.getOutput(inp);

    isDefined(out);

    const spendTx = await blaze
      .newTransaction()
      .addInput(new TransactionUnspentOutput(inp, out), VOID_PLUTUS_DATA)
      .addReferenceInput(refUtxo)
      .complete();
    const spendTxHash = await signAndSubmit(spendTx, blaze);
    emulator.awaitTransactionConfirmation(spendTxHash);
    const out2 = emulator.getOutput(new TransactionInput(spendTxHash, 0n));
    isDefined(out2);
    expect(out2.address()).toEqual(wallet1.address);
  });

  test("Should be able to spend a UTxO with a self-reference script", async () => {
    const addr = addressFromCredential(
      NetworkId.Testnet,
      Credential.fromCore({
        type: CredentialType.ScriptHash,
        hash: alwaysTrueScript.hash(),
      }),
    );

    const tx = await blaze
      .newTransaction()
      .lockAssets(
        addr,
        makeValue(1_000_000_000n),
        ONE_PLUTUS_DATA,
        alwaysTrueScript,
      )
      .complete();

    const txHash = await signAndSubmit(tx, blaze);
    emulator.awaitTransactionConfirmation(txHash);
    const inp = new TransactionInput(txHash, 0n);
    const out = emulator.getOutput(inp);

    isDefined(out);

    const spendTx = await blaze
      .newTransaction()
      .addInput(new TransactionUnspentOutput(inp, out), VOID_PLUTUS_DATA)
      .complete();
    const spendTxHash = await signAndSubmit(spendTx, blaze);
    emulator.awaitTransactionConfirmation(spendTxHash);
    const out2 = emulator.getOutput(new TransactionInput(spendTxHash, 0n));
    isDefined(out2);
    expect(out2.address()).toEqual(wallet1.address);
  });

  test("Should be able to mint from a policy", async () => {
    const policy = PolicyId(alwaysTrueScript.hash());
    const tx = await blaze
      .newTransaction()
      .addMint(policy, new Map([[AssetName(""), 1n]]), VOID_PLUTUS_DATA)
      .provideScript(alwaysTrueScript)
      .complete();
    const txHash = await signAndSubmit(tx, blaze);
    emulator.awaitTransactionConfirmation(txHash);
    const out = emulator.getOutput(new TransactionInput(txHash, 0n));
    isDefined(out);
    expect(out.amount().multiasset()?.get(AssetId(policy))).toEqual(1n);
  });

  test("Should be able to register stake for a script", async () => {
    const cred = {
      hash: alwaysTrueScript.hash(),
      type: CredentialType.ScriptHash,
    };
    const rewardAccount = RewardAccount.fromCredential(cred, NetworkId.Testnet);
    expect(emulator.accounts).not.toHaveProperty(rewardAccount);

    const tx = await blaze
      .newTransaction()
      .addRegisterStake(Credential.fromCore(cred))
      .complete();

    const txHash = await signAndSubmit(tx, blaze);
    emulator.awaitTransactionConfirmation(txHash);

    expect(emulator.accounts.get(rewardAccount)).toEqual(0n);
  });

  test("Should be able to withdraw from a script", async () => {
    const rewardAccount = RewardAccount.fromCredential(
      {
        hash: alwaysTrueScript.hash(),
        type: CredentialType.ScriptHash,
      },
      NetworkId.Testnet,
    );

    emulator.accounts.set(rewardAccount, 10_000_000n);
    const tx = await blaze
      .newTransaction()
      .addWithdrawal(rewardAccount, 10_000_000n, VOID_PLUTUS_DATA)
      .provideScript(alwaysTrueScript)
      // TODO: fix coin selection so you don't have to do this
      .addInput((await wallet1.getUnspentOutputs())[0]!)
      .complete();

    const txHash = await signAndSubmit(tx, blaze);
    emulator.awaitTransactionConfirmation(txHash);
    const out = emulator.getOutput(new TransactionInput(txHash, 0n));
    isDefined(out);
    expect(out.address()).toEqual(wallet1.address);
  });
});
