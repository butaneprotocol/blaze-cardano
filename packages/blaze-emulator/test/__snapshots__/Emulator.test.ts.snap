// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emulator Should not be able to spend if inputs/outputs are not balanced 1`] = `"Value not conserved. Leftover Value: -1342617, "`;
